File: go/schemasource/schemasource.go
// SPDX-License-Identifier: Apache-2.0

// Package schemasource resolves a "schema source reference" into a
// Keystone desired-state SchemaDefinitionSpec, normalising every input
// through a common representation.
//
// A source reference names where a desired (or current) schema lives:
//
//	yaml://path/to/schema-definition.yaml   declarative schema-as-code
//	sql://path/to/schema.sql                raw DDL (hand-written, ORM
//	file://path/to/schema.sql               provider output, EF Core
//	                                        GenerateCreateScript, pg_dump)
//	db://postgres://user:pw@host/db?schema=public
//	                                        an existing database already
//	                                        holding the desired shape
//	exec://dotnet keystone-ef               a provider program that prints
//	                                        the desired schema as DDL on
//	                                        stdout (any ORM, any runtime)
//
// The unifying idea — mirroring Atlas's "dev database" — is that a
// throwaway PostgreSQL schema is the universal normaliser. SQL DDL from
// any origin (including any ORM that can emit CREATE statements) is
// applied to a scratch schema on a dev database and read back with the
// same drift.Inspector that reads production. That is why Keystone
// supports "schema from code" for both declarative SQL/YAML and ORM
// entities without re-implementing a single ORM's model layer: the ORM
// emits SQL, the dev database turns SQL into a Snapshot, and the existing
// differ takes it from there.
package schemasource

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net/url"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"sigs.k8s.io/yaml"

	"github.com/dogukanturhal/keystone-sdk/go/drift"
	"github.com/dogukanturhal/keystone-sdk/go/migration"
	"github.com/dogukanturhal/keystone-sdk/go/schemaspec"
	pg "github.com/dogukanturhal/keystone-sdk/go/sqlquote"
	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Scheme identifies the kind of a parsed source reference.
type Scheme string

const (
	// SchemeYAML is a SchemaDefinition YAML file (declarative).
	SchemeYAML Scheme = "yaml"
	// SchemeSQL is a raw DDL file, normalised through a dev database.
	SchemeSQL Scheme = "sql"
	// SchemeDB is a live database inspected directly.
	SchemeDB Scheme = "db"
	// SchemeExec is a provider program whose stdout is DDL, normalised
	// through a dev database exactly like SchemeSQL. This is the ORM
	// integration seam: any ORM that can render CREATE statements — EF
	// Core, GORM, Prisma, SQLAlchemy, Django, Ecto — plugs in without
	// Keystone learning anything about its model layer.
	//
	// Running a program is a capability, not a parse result: it is
	// refused unless the caller sets Resolver.AllowExec. See Resolver.
	SchemeExec Scheme = "exec"
)

// Ref is a parsed source reference.
type Ref struct {
	Scheme Scheme
	// Path is the file path for yaml:// and sql:// references.
	Path string
	// DSN is the PostgreSQL connection string for db:// references.
	DSN string
	// Schema is the schema to inspect for db:// references when carried
	// in the reference (?schema=). Empty means the caller's default.
	Schema string
	// Program is the argv of an exec:// provider — never a shell string.
	// Populated by ParseRef from the reference text, or set directly by
	// callers that already hold a structured array (keystone.yaml).
	Program []string
}

// ParseRef parses a source reference string. Bare paths (no scheme) are
// classified by extension: .sql → SchemeSQL, .yaml/.yml → SchemeYAML.
func ParseRef(s string) (Ref, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return Ref{}, fmt.Errorf("empty source reference")
	}
	switch {
	case strings.HasPrefix(s, "yaml://"):
		return Ref{Scheme: SchemeYAML, Path: strings.TrimPrefix(s, "yaml://")}, nil
	case strings.HasPrefix(s, "sql://"):
		return Ref{Scheme: SchemeSQL, Path: strings.TrimPrefix(s, "sql://")}, nil
	case strings.HasPrefix(s, "file://"):
		p := strings.TrimPrefix(s, "file://")
		if strings.HasSuffix(strings.ToLower(p), ".sql") {
			return Ref{Scheme: SchemeSQL, Path: p}, nil
		}
		return Ref{Scheme: SchemeYAML, Path: p}, nil
	case strings.HasPrefix(s, "exec://"):
		program, err := SplitProgram(strings.TrimPrefix(s, "exec://"))
		if err != nil {
			return Ref{}, err
		}
		return Ref{Scheme: SchemeExec, Program: program}, nil
	case strings.HasPrefix(s, "db://"):
		dsn := strings.TrimPrefix(s, "db://")
		// Allow db://<bare> as shorthand for a postgres:// DSN.
		if !strings.HasPrefix(dsn, "postgres://") && !strings.HasPrefix(dsn, "postgresql://") {
			dsn = "postgres://" + dsn
		}
		schema, clean, err := extractSchemaParam(dsn)
		if err != nil {
			return Ref{}, err
		}
		return Ref{Scheme: SchemeDB, DSN: clean, Schema: schema}, nil
	default:
		// Bare path: classify by extension.
		low := strings.ToLower(s)
		switch {
		case strings.HasSuffix(low, ".sql"):
			return Ref{Scheme: SchemeSQL, Path: s}, nil
		case strings.HasSuffix(low, ".yaml"), strings.HasSuffix(low, ".yml"):
			return Ref{Scheme: SchemeYAML, Path: s}, nil
		default:
			return Ref{}, fmt.Errorf("cannot classify source reference %q: prefix with yaml:// , sql:// , db:// or exec:// (or use a .sql/.yaml path)", s)
		}
	}
}

// extractSchemaParam pulls a `schema` query parameter out of a DSN and
// returns the schema plus the DSN with that parameter removed (so it is
// never passed to PostgreSQL, which would reject an unknown parameter).
func extractSchemaParam(dsn string) (schema, clean string, err error) {
	u, perr := url.Parse(dsn)
	if perr != nil {
		// Not a URL-style DSN (could be a keyword/value string); leave as-is.
		return "", dsn, nil //nolint:nilerr // keyword DSNs carry no schema param
	}
	q := u.Query()
	schema = q.Get("schema")
	if schema != "" {
		q.Del("schema")
		u.RawQuery = q.Encode()
	}
	return schema, u.String(), nil
}

// Resolver resolves source references with an explicit capability set.
//
// The zero Resolver is the safe one: it behaves exactly like the
// package-level ResolveDesired and refuses exec://. Running a program
// named by a reference is a privilege that only a developer-driven
// caller (keystonectl, a CI job) may grant. A reconciler resolving a
// reference that arrived in a CR spec must never set AllowExec — that
// would turn a namespace-scoped write into remote code execution inside
// the operator pod.
type Resolver struct {
	// AllowExec enables the exec:// scheme. Off by default, on purpose.
	AllowExec bool
	// ExecDir is the working directory for exec:// providers. Empty
	// means the current process's directory. Set this to the project
	// root so a provider resolves its ORM project relative to the
	// config file rather than to wherever the CLI was invoked.
	ExecDir string
	// ExecEnv is the environment for exec:// providers. Nil inherits the
	// parent environment — which is what an ORM provider normally needs
	// (PATH, DOTNET_ROOT, GOCACHE, connection strings).
	ExecEnv []string
	// ExecTimeout bounds a provider's runtime. Zero means
	// DefaultExecTimeout.
	ExecTimeout time.Duration
}

// ResolveDesired resolves ref into a desired-state SchemaDefinitionSpec
// with exec:// disabled. Equivalent to Resolver{}.ResolveDesired.
func ResolveDesired(ctx context.Context, ref Ref, devPool *pgxpool.Pool, defaultSchema string, opts schemaspec.Options) (*keystonev1alpha1.SchemaDefinitionSpec, error) {
	return Resolver{}.ResolveDesired(ctx, ref, devPool, defaultSchema, opts)
}

// ResolveDesired resolves ref into a desired-state SchemaDefinitionSpec.
//
//   - yaml:// reads the SchemaDefinition file (devPool unused).
//   - sql:// applies the DDL to a scratch schema on devPool, inspects it,
//     and converts the resulting Snapshot (devPool is REQUIRED).
//   - db:// opens its own connection to ref.DSN, inspects ref.Schema (or
//     defaultSchema when the ref carries none), and converts.
//
// opts is forwarded to schemaspec.FromSnapshot for the sql:// / db://
// paths (schemaRef / selector labels); it is ignored for yaml:// where
// the file already carries those fields.
func (r Resolver) ResolveDesired(ctx context.Context, ref Ref, devPool *pgxpool.Pool, defaultSchema string, opts schemaspec.Options) (*keystonev1alpha1.SchemaDefinitionSpec, error) {
	switch ref.Scheme {
	case SchemeYAML:
		return LoadSpecFile(ref.Path)
	case SchemeSQL:
		if devPool == nil {
			return nil, fmt.Errorf("sql:// source requires a --dev-url database to normalise the DDL")
		}
		ddl, err := os.ReadFile(ref.Path)
		if err != nil {
			return nil, fmt.Errorf("read %s: %w", ref.Path, err)
		}
		snap, err := SnapshotFromDDL(ctx, devPool, string(ddl), defaultSchema)
		if err != nil {
			return nil, err
		}
		return schemaspec.FromSnapshot(snap, opts), nil
	case SchemeExec:
		if !r.AllowExec {
			return nil, fmt.Errorf("exec:// sources are disabled for this caller: running a provider program is opt-in (schemasource.Resolver.AllowExec)")
		}
		if devPool == nil {
			return nil, fmt.Errorf("exec:// source requires a --dev-url database to normalise the DDL")
		}
		ddl, err := runProgram(ctx, ref.Program, r.ExecDir, r.ExecEnv, r.ExecTimeout)
		if err != nil {
			return nil, err
		}
		snap, err := SnapshotFromDDL(ctx, devPool, ddl, defaultSchema)
		if err != nil {
			return nil, fmt.Errorf("normalise DDL from exec:// provider %q: %w",
				strings.Join(ref.Program, " "), err)
		}
		return schemaspec.FromSnapshot(snap, opts), nil
	case SchemeDB:
		schema := ref.Schema
		if schema == "" {
			schema = defaultSchema
		}
		pool, err := pgxpool.New(ctx, ref.DSN)
		if err != nil {
			return nil, fmt.Errorf("connect to db:// source: %w", err)
		}
		defer pool.Close()
		snap, err := drift.NewInspector(pool).Inspect(ctx, schema)
		if err != nil {
			return nil, fmt.Errorf("inspect db:// source schema %q: %w", schema, err)
		}
		return schemaspec.FromSnapshot(snap, opts), nil
	default:
		return nil, fmt.Errorf("unsupported source scheme %q", ref.Scheme)
	}
}

// LoadSpecFile reads a SchemaDefinition YAML — accepting either a full CR
// (apiVersion/kind/spec) or a bare spec document — into a spec.
func LoadSpecFile(path string) (*keystonev1alpha1.SchemaDefinitionSpec, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", path, err)
	}
	var full struct {
		Spec keystonev1alpha1.SchemaDefinitionSpec `yaml:"spec" json:"spec"`
	}
	if err := yaml.Unmarshal(raw, &full); err == nil && len(full.Spec.Tables) > 0 {
		return &full.Spec, nil
	}
	var spec keystonev1alpha1.SchemaDefinitionSpec
	if err := yaml.Unmarshal(raw, &spec); err != nil {
		return nil, fmt.Errorf("unmarshal %s: %w", path, err)
	}
	return &spec, nil
}

// SnapshotFromDDL applies a DDL blob to a fresh, isolated scratch schema
// on devPool, inspects it, and returns the Snapshot — the canonical "dev
// database" normalisation. The scratch schema is created and dropped
// within the call. The DDL must be schema-relative (unqualified object
// names); the runner sets search_path to the scratch schema so objects
// land there.
//
// targetSchema is the logical schema the resulting snapshot should report
// (its .Schema and the schema qualifier embedded in inspected DDL such as
// index definitions). Pass the schema the diff will run against so the
// differ's index/object comparison is not confused by the throwaway
// scratch schema name.
func SnapshotFromDDL(ctx context.Context, devPool *pgxpool.Pool, ddl, targetSchema string) (*drift.Snapshot, error) {
	return materialize(ctx, devPool, map[string]string{"desired.up.sql": ddl}, []string{"desired.up.sql"}, targetSchema)
}

// SnapshotFromFiles replays an ordered set of migration files onto a
// scratch schema on devPool and inspects the result — the dev-replay
// basis for `migrate diff`'s current state. files maps name→content; the
// apply order is lexicographic by name (mirroring the runner). Pass the
// directory's *.up.sql files; down files must be excluded by the caller.
// targetSchema has the same meaning as in SnapshotFromDDL.
func SnapshotFromFiles(ctx context.Context, devPool *pgxpool.Pool, files map[string]string, targetSchema string) (*drift.Snapshot, error) {
	names := make([]string, 0, len(files))
	for n := range files {
		names = append(names, n)
	}
	sort.Strings(names)
	return materialize(ctx, devPool, files, names, targetSchema)
}

// materialize is the shared dev-database normaliser: create scratch
// schema → replay files via the production runner (which handles
// search_path, statement splitting, and CONCURRENTLY/tx mode) → inspect
// → drop scratch schema. Reusing migration.Runner means the dev replay is
// byte-for-byte the same apply path production uses, so the resulting
// Snapshot is high-fidelity.
func materialize(ctx context.Context, devPool *pgxpool.Pool, files map[string]string, names []string, targetSchema string) (snap *drift.Snapshot, err error) {
	if devPool == nil {
		return nil, fmt.Errorf("materialize: nil dev pool")
	}
	scratch, err := scratchSchemaName()
	if err != nil {
		return nil, err
	}
	if _, err := devPool.Exec(ctx, "CREATE SCHEMA "+pg.QuoteIdentifier(scratch)); err != nil {
		return nil, fmt.Errorf("create scratch schema: %w", err)
	}
	// Always tear the scratch schema down, even on failure. Use a
	// background context so cleanup runs even if ctx is already cancelled.
	defer func() {
		_, derr := devPool.Exec(context.Background(),
			"DROP SCHEMA IF EXISTS "+pg.QuoteIdentifier(scratch)+" CASCADE")
		if derr != nil && err == nil {
			err = fmt.Errorf("drop scratch schema %q: %w", scratch, derr)
		}
	}()

	runner, err := migration.NewRunner(devPool, scratch, "keystone_schema_migrations", "")
	if err != nil {
		return nil, fmt.Errorf("new runner: %w", err)
	}
	if err := runner.EnsureBookkeeping(ctx); err != nil {
		return nil, fmt.Errorf("ensure bookkeeping: %w", err)
	}
	src := &migration.ResolvedSource{Files: files, Names: names}
	src.ContentHash = migration.HashFiles(src.Files, src.Names)
	if _, err := runner.Apply(ctx, "materialize", src.ContentHash, src); err != nil {
		return nil, fmt.Errorf("apply DDL to scratch schema: %w", err)
	}

	snap, err = drift.NewInspector(devPool).Inspect(ctx, scratch)
	if err != nil {
		return nil, fmt.Errorf("inspect scratch schema: %w", err)
	}
	// The inspected snapshot embeds the throwaway scratch schema name in
	// its .Schema and in every catalog-derived DDL string (index/view/
	// function/policy definitions). Rewrite it to the caller's target
	// schema so the differ — which compares observed index DDL against
	// DDL it renders with the target schema — does not see the scratch
	// name as a spurious difference. The scratch name is a unique random
	// token, so a literal replacement cannot collide with real content.
	if targetSchema != "" && targetSchema != scratch {
		rewriteSnapshotSchema(snap, scratch, targetSchema)
	}
	return snap, nil
}

// rewriteSnapshotSchema replaces every occurrence of the schema name
// `from` with `to` across a snapshot's schema-qualified DDL strings.
// Safe because `from` is the unique scratch-schema token.
func rewriteSnapshotSchema(snap *drift.Snapshot, from, to string) {
	repl := func(s string) string { return strings.ReplaceAll(s, from, to) }
	snap.Schema = to

	// Extensions created by the desired source land in the scratch schema, so
	// their recorded namespace is the throwaway token. Left unrewritten it is
	// written verbatim into the authored migration as
	// CREATE EXTENSION ... SCHEMA "keystone_dev_<hash>".
	for i := range snap.Extensions {
		snap.Extensions[i].Schema = repl(snap.Extensions[i].Schema)
	}

	// format_type qualifies user-defined types with their schema, which is the
	// scratch name for anything the desired source created.
	for i := range snap.Tables {
		for j := range snap.Tables[i].Columns {
			snap.Tables[i].Columns[j].FormattedType = repl(snap.Tables[i].Columns[j].FormattedType)

			// A generation expression is stored fully qualified by PostgreSQL, so
			// every function and cast in it carries the scratch schema name. Here the
			// qualifier is REMOVED rather than rewritten to the target schema.
			//
			// Rewriting would bake in where the extension happened to live when the
			// migration was authored. Replaying that migration installs the extension
			// into a fresh scratch schema, and a hard-coded public.st_setsrid then
			// fails with 42704 type "public.geography" does not exist. Unqualified, it
			// resolves through the runner's search_path (target schema, then public)
			// in both cases.
			snap.Tables[i].Columns[j].Generated = strings.ReplaceAll(
				snap.Tables[i].Columns[j].Generated, from+".", "")

			// A DEFAULT is stored qualified for exactly the same reasons:
			// nextval('<schema>.seq'::regclass) for a serial-style column,
			// and '<label>'::<schema>.<enum> for an enum default. Left
			// alone, the scratch token is written straight into the
			// authored migration and the apply fails at the target with
			// 3F000 schema "keystone_dev_<hash>" does not exist — the
			// migration is broken before it is ever reviewed.
			//
			// Stripped rather than rewritten, matching Generated above: an
			// unqualified reference resolves through the runner's
			// search_path, so one bundle stays applicable to every schema
			// it fans out to instead of being pinned to whichever schema
			// happened to author it.
			snap.Tables[i].Columns[j].Default = strings.ReplaceAll(
				snap.Tables[i].Columns[j].Default, from+".", "")
		}
	}
	for i := range snap.Indexes {
		snap.Indexes[i].Definition = repl(snap.Indexes[i].Definition)
	}
	for i := range snap.Constraints {
		snap.Constraints[i].Definition = repl(snap.Constraints[i].Definition)
	}
	for i := range snap.Tables {
		snap.Tables[i].ViewDefinition = repl(snap.Tables[i].ViewDefinition)
	}
	for i := range snap.Functions {
		snap.Functions[i].Definition = repl(snap.Functions[i].Definition)
	}
	for i := range snap.Policies {
		snap.Policies[i].Using = repl(snap.Policies[i].Using)
		snap.Policies[i].WithCheck = repl(snap.Policies[i].WithCheck)
	}
	for i := range snap.MaterializedViews {
		snap.MaterializedViews[i].Definition = repl(snap.MaterializedViews[i].Definition)
	}
}

// scratchSchemaName returns a unique, identifier-valid scratch schema
// name. The random suffix avoids collisions between concurrent runs
// sharing one dev database.
func scratchSchemaName() (string, error) {
	var b [6]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("scratch schema name: %w", err)
	}
	return "keystone_dev_" + hex.EncodeToString(b[:]), nil
}

File: rust/keystone-sdk/src/schemasource.rs
// SPDX-License-Identifier: Apache-2.0

//! Resolves a "schema source reference" into a desired-state
//! [`SchemaDefinitionSpec`], normalising every input through a common
//! representation.
//!
//! Ported from the Go `schemasource` package. Reference forms:
//! `yaml://path` (declarative spec), `sql://path` / `file://*.sql` (raw DDL
//! normalised through a dev database), `db://<dsn>?schema=…` (inspect a live
//! database). The unifying idea — Atlas's "dev database" — is that a
//! throwaway scratch schema is the universal normaliser: DDL from any origin
//! is applied to a scratch schema via the production runner and read back by
//! the same inspector, so [`from_snapshot`] + the differ take it from there.

use std::collections::BTreeMap;

use tokio_postgres::Client;

use crate::declarative::spec::SchemaDefinitionSpec;
use crate::drift::{Inspector, Snapshot};
use crate::ident::quote_identifier;
use crate::runner::Runner;
use crate::schemaspec::from_snapshot;
use crate::source::{resolve, SqlFile};
use crate::tracking;

/// The kind of a parsed source reference.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Scheme {
    /// A SchemaDefinition YAML file (declarative).
    Yaml,
    /// A raw DDL file, normalised through a dev database.
    Sql,
    /// A live database inspected directly.
    Db,
}

/// A parsed source reference.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Ref {
    pub scheme: Scheme,
    /// File path for `yaml://` / `sql://`.
    pub path: String,
    /// PostgreSQL connection string for `db://`.
    pub dsn: String,
    /// Schema to inspect for `db://` (from `?schema=`); empty = caller default.
    pub schema: String,
}

/// Errors from this package.
#[derive(Debug, thiserror::Error)]
pub enum SchemaSourceError {
    #[error("{0}")]
    Parse(String),
    #[error("{0}")]
    Io(String),
    #[error("unmarshal spec: {0}")]
    Yaml(String),
    #[error("sql:// source requires a --dev-url database to normalise the DDL")]
    DevDbRequired,
    #[error(transparent)]
    Ident(#[from] crate::ident::IdentError),
    #[error(transparent)]
    Db(tokio_postgres::Error),
    #[error(transparent)]
    Connect(#[from] crate::pgconn::ConnectError),
    #[error(transparent)]
    Runner(#[from] crate::runner::RunnerError),
    #[error(transparent)]
    Inspect(#[from] crate::drift::inspector::InspectError),
}

/// Parses a source reference string. Bare paths (no scheme) are classified
/// by extension: `.sql` → SQL, `.yaml`/`.yml` → YAML.
pub fn parse_ref(s: &str) -> Result<Ref, SchemaSourceError> {
    let s = s.trim();
    if s.is_empty() {
        return Err(SchemaSourceError::Parse("empty source reference".into()));
    }
    if let Some(p) = s.strip_prefix("yaml://") {
        return Ok(Ref {
            scheme: Scheme::Yaml,
            path: p.to_string(),
            dsn: String::new(),
            schema: String::new(),
        });
    }
    if let Some(p) = s.strip_prefix("sql://") {
        return Ok(Ref {
            scheme: Scheme::Sql,
            path: p.to_string(),
            dsn: String::new(),
            schema: String::new(),
        });
    }
    if let Some(p) = s.strip_prefix("file://") {
        let scheme = if p.to_lowercase().ends_with(".sql") {
            Scheme::Sql
        } else {
            Scheme::Yaml
        };
        return Ok(Ref {
            scheme,
            path: p.to_string(),
            dsn: String::new(),
            schema: String::new(),
        });
    }
    if let Some(d) = s.strip_prefix("db://") {
        let mut dsn = d.to_string();
        if !dsn.starts_with("postgres://") && !dsn.starts_with("postgresql://") {
            dsn = format!("postgres://{dsn}");
        }
        let (schema, clean) = extract_schema_param(&dsn);
        return Ok(Ref {
            scheme: Scheme::Db,
            path: String::new(),
            dsn: clean,
            schema,
        });
    }
    let low = s.to_lowercase();
    if low.ends_with(".sql") {
        return Ok(Ref {
            scheme: Scheme::Sql,
            path: s.to_string(),
            dsn: String::new(),
            schema: String::new(),
        });
    }
    if low.ends_with(".yaml") || low.ends_with(".yml") {
        return Ok(Ref {
            scheme: Scheme::Yaml,
            path: s.to_string(),
            dsn: String::new(),
            schema: String::new(),
        });
    }
    Err(SchemaSourceError::Parse(format!(
        "cannot classify source reference {s:?}: prefix with yaml:// , sql:// , or db:// (or use a .sql/.yaml path)"
    )))
}

/// Pulls a `schema` query parameter out of a DSN, returning the schema plus
/// the DSN with that parameter removed (so PostgreSQL never sees it).
fn extract_schema_param(dsn: &str) -> (String, String) {
    let Some(qpos) = dsn.find('?') else {
        return (String::new(), dsn.to_string());
    };
    let (base, query) = dsn.split_at(qpos);
    let query = &query[1..]; // after '?'
    let mut schema = String::new();
    let kept: Vec<&str> = query
        .split('&')
        .filter(|kv| {
            if let Some(v) = kv.strip_prefix("schema=") {
                schema = v.to_string();
                false
            } else {
                !kv.is_empty()
            }
        })
        .collect();
    let clean = if kept.is_empty() {
        base.to_string()
    } else {
        format!("{base}?{}", kept.join("&"))
    };
    (schema, clean)
}

/// Parses a SchemaDefinition YAML string — accepting either a full CR
/// (`spec:` key) or a bare spec document — into a [`SchemaDefinitionSpec`].
pub fn load_spec_str(raw: &str) -> Result<SchemaDefinitionSpec, SchemaSourceError> {
    #[derive(serde::Deserialize, Default)]
    #[serde(default)]
    struct Wrapper {
        spec: SchemaDefinitionSpec,
    }
    if let Ok(w) = serde_yaml::from_str::<Wrapper>(raw) {
        if !w.spec.tables.is_empty() {
            return Ok(w.spec);
        }
    }
    serde_yaml::from_str(raw).map_err(|e| SchemaSourceError::Yaml(e.to_string()))
}

/// Reads a SchemaDefinition YAML file into a spec.
pub fn load_spec_file(path: &str) -> Result<SchemaDefinitionSpec, SchemaSourceError> {
    let raw = std::fs::read_to_string(path)
        .map_err(|e| SchemaSourceError::Io(format!("read {path}: {e}")))?;
    load_spec_str(&raw)
}

/// Resolves `reference` into a desired-state spec.
///
/// - `yaml://` reads the file (`dev_client` unused).
/// - `sql://` applies the DDL to a scratch schema on `dev_client` (REQUIRED),
///   inspects it, and converts.
/// - `db://` opens its own connection to `reference.dsn`, inspects
///   `reference.schema` (or `default_schema`), and converts.
pub async fn resolve_desired(
    reference: &Ref,
    dev_client: Option<&mut Client>,
    default_schema: &str,
) -> Result<SchemaDefinitionSpec, SchemaSourceError> {
    match reference.scheme {
        Scheme::Yaml => load_spec_file(&reference.path),
        Scheme::Sql => {
            let dev = dev_client.ok_or(SchemaSourceError::DevDbRequired)?;
            let ddl = std::fs::read_to_string(&reference.path)
                .map_err(|e| SchemaSourceError::Io(format!("read {}: {e}", reference.path)))?;
            let snap = snapshot_from_ddl(dev, &ddl, default_schema).await?;
            Ok(from_snapshot(&snap))
        }
        Scheme::Db => {
            let schema = if reference.schema.is_empty() {
                default_schema
            } else {
                &reference.schema
            };
            let (client, conn_task) = crate::pgconn::connect(&reference.dsn)
                .await
                .map_err(SchemaSourceError::Connect)?;
            let snap = Inspector::new().inspect(&client, schema).await;
            drop(client); // ends the connection so the spawned task finishes
            let _ = conn_task.await;
            Ok(from_snapshot(&snap?))
        }
    }
}

/// Applies a DDL blob to a fresh scratch schema on `client`, inspects it, and
/// returns the Snapshot (the dev-database normalisation). `target_schema` is
/// the logical schema the snapshot should report (its `.schema` and the
/// schema qualifier in inspected DDL).
pub async fn snapshot_from_ddl(
    client: &mut Client,
    ddl: &str,
    target_schema: &str,
) -> Result<Snapshot, SchemaSourceError> {
    materialize(
        client,
        vec![SqlFile {
            name: "desired.up.sql".into(),
            body: ddl.to_string(),
        }],
        target_schema,
    )
    .await
}

/// Replays an ordered set of migration files onto a scratch schema and
/// inspects the result. Apply order is lexicographic by name (mirroring the
/// runner). Pass `*.up.sql` files only.
pub async fn snapshot_from_files(
    client: &mut Client,
    files: BTreeMap<String, String>,
    target_schema: &str,
) -> Result<Snapshot, SchemaSourceError> {
    // BTreeMap iterates sorted by name → SqlFile vec in lexicographic order.
    let v: Vec<SqlFile> = files
        .into_iter()
        .map(|(name, body)| SqlFile { name, body })
        .collect();
    materialize(client, v, target_schema).await
}

/// The shared dev-database normaliser: create scratch schema → replay files
/// via the production runner → inspect → drop scratch schema (always, even on
/// error). Reusing [`Runner`] means the dev replay is the exact apply path
/// production uses, so the snapshot is high-fidelity.
async fn materialize(
    client: &mut Client,
    files: Vec<SqlFile>,
    target_schema: &str,
) -> Result<Snapshot, SchemaSourceError> {
    let scratch = scratch_schema_name();
    client
        .batch_execute(&format!("CREATE SCHEMA {}", quote_identifier(&scratch)))
        .await
        .map_err(SchemaSourceError::Db)?;

    let result = materialize_inner(client, &scratch, files, target_schema).await;

    // Always tear the scratch schema down.
    let drop_res = client
        .batch_execute(&format!(
            "DROP SCHEMA IF EXISTS {} CASCADE",
            quote_identifier(&scratch)
        ))
        .await;

    match (result, drop_res) {
        (Ok(snap), Ok(_)) => Ok(snap),
        (Ok(_), Err(e)) => Err(SchemaSourceError::Db(e)),
        (Err(e), _) => Err(e),
    }
}

async fn materialize_inner(
    client: &mut Client,
    scratch: &str,
    files: Vec<SqlFile>,
    target_schema: &str,
) -> Result<Snapshot, SchemaSourceError> {
    let runner = Runner::new(scratch, "keystone_schema_migrations", None)?;
    tracking::ensure_bookkeeping(client, scratch, "keystone_schema_migrations")
        .await
        .map_err(SchemaSourceError::Db)?;

    let src = resolve(files);
    let content_hash = src.content_hash().to_string();
    runner
        .apply(client, "materialize", &content_hash, &src)
        .await?;

    let mut snap = Inspector::new().inspect(client, scratch).await?;
    if !target_schema.is_empty() && target_schema != scratch {
        rewrite_snapshot_schema(&mut snap, scratch, target_schema);
    }
    Ok(snap)
}

/// Replaces every occurrence of schema name `from` with `to` across the
/// snapshot's schema-qualified DDL strings. Safe because `from` is the unique
/// scratch-schema token.
fn rewrite_snapshot_schema(snap: &mut Snapshot, from: &str, to: &str) {
    snap.schema = to.to_string();
    for idx in &mut snap.indexes {
        idx.definition = idx.definition.replace(from, to);
    }
    for c in &mut snap.constraints {
        c.definition = c.definition.replace(from, to);
    }
    for t in &mut snap.tables {
        t.view_definition = t.view_definition.replace(from, to);
    }
    for f in &mut snap.functions {
        f.definition = f.definition.replace(from, to);
    }
    for p in &mut snap.policies {
        p.using = p.using.replace(from, to);
        p.with_check = p.with_check.replace(from, to);
    }
    for mv in &mut snap.materialized_views {
        mv.definition = mv.definition.replace(from, to);
    }
}

/// A unique, identifier-valid scratch schema name (process id + monotonic
/// counter + time, avoiding collisions between concurrent runs).
fn scratch_schema_name() -> String {
    use std::sync::atomic::{AtomicU64, Ordering};
    static COUNTER: AtomicU64 = AtomicU64::new(0);
    let c = COUNTER.fetch_add(1, Ordering::Relaxed);
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(0);
    format!("keystone_dev_{:x}{:x}{:x}", std::process::id(), nanos, c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_ref_schemes() {
        assert_eq!(parse_ref("yaml://a/b.yaml").unwrap().scheme, Scheme::Yaml);
        assert_eq!(parse_ref("sql://a/b.sql").unwrap().scheme, Scheme::Sql);
        assert_eq!(parse_ref("file://x.sql").unwrap().scheme, Scheme::Sql);
        assert_eq!(parse_ref("file://x.yaml").unwrap().scheme, Scheme::Yaml);
        // Bare-path classification by extension.
        assert_eq!(parse_ref("schema.sql").unwrap().scheme, Scheme::Sql);
        assert_eq!(parse_ref("def.yml").unwrap().scheme, Scheme::Yaml);
        // Unclassifiable.
        assert!(parse_ref("nope.txt").is_err());
        assert!(parse_ref("  ").is_err());
    }

    #[test]
    fn parse_ref_db_dsn() {
        let r = parse_ref("db://user:pw@host/db?schema=public&sslmode=disable").unwrap();
        assert_eq!(r.scheme, Scheme::Db);
        assert_eq!(r.schema, "public");
        // `schema` removed; `postgres://` prepended; other params kept.
        assert_eq!(r.dsn, "postgres://user:pw@host/db?sslmode=disable");

        // schema is the only param → '?' dropped.
        let r = parse_ref("db://postgres://h/db?schema=app").unwrap();
        assert_eq!(r.schema, "app");
        assert_eq!(r.dsn, "postgres://h/db");

        // No schema param.
        let r = parse_ref("db://postgresql://h/db").unwrap();
        assert_eq!(r.schema, "");
        assert_eq!(r.dsn, "postgresql://h/db");
    }

    #[test]
    fn load_spec_str_bare_and_wrapped() {
        let bare = "tables:\n  - name: users\n    columns:\n      - name: id\n        type: bigint\n        nullable: false\n        primaryKey: true\n";
        let spec = load_spec_str(bare).unwrap();
        assert_eq!(spec.tables.len(), 1);
        assert_eq!(spec.tables[0].name, "users");
        assert!(spec.tables[0].columns[0].primary_key);
        assert_eq!(spec.tables[0].columns[0].r#type, "bigint");

        let wrapped = "apiVersion: keystone.hexxlock.io/v1alpha1\nkind: SchemaDefinition\nmetadata:\n  name: x\nspec:\n  allowDestructive: true\n  tables:\n    - name: t\n      columns:\n        - name: id\n          type: int\n";
        let spec = load_spec_str(wrapped).unwrap();
        assert_eq!(spec.tables.len(), 1);
        assert!(spec.allow_destructive);
        assert_eq!(spec.tables[0].name, "t");
    }

    #[test]
    fn load_spec_str_camelcase_fields() {
        // enableRLS + where + columnRefs round-trip through the camelCase tags.
        let y = "tables:\n  - name: t\n    enableRLS: true\n    columns:\n      - name: id\n        type: int\n    indexes:\n      - name: ix\n        where: deleted_at IS NULL\n        columnRefs:\n          - name: a\n            direction: desc\n";
        let spec = load_spec_str(y).unwrap();
        assert!(spec.tables[0].enable_rls);
        assert_eq!(spec.tables[0].indexes[0].where_, "deleted_at IS NULL");
        assert_eq!(spec.tables[0].indexes[0].column_refs[0].direction, "desc");
    }
}
